[WIP] Memory channels

sabtf opened the pull request.

sabtf committed:
Add memory channels with fixed address ranges to ProcessContext

sabtf committed:
Add memory channel for single Trustlet use

mmisono commented on kernel/src/process_manager/process.rs:
```diff
@@ -171,9 +173,14 @@ impl TrustedProcess {
         let trustlet = TrustedProcess::dublicate(parent);
+        if data != 0 {
+            let (function_code, function_code_range) = ProcessPageTableRef::copy_data_from_guest(data, size, pgt);
+            trustlet.base.page_table_ref.add_function(function_code, size);
```
`page_table_ref.add_function()` (more precicely, `add_region_vaddr()` expects page-aligned size, therefore
```suggestion
            let size = (4096 - (size & 0xFFF)) + size;
            trustlet.base.page_table_ref.add_function(function_code, size);
```

(I think `add_region_vaddr` should be able to take non-page aligend size though)

mmisono commented on kernel/src/process_manager/process_paging.rs:
```diff
@@ -286,6 +286,12 @@ impl ProcessPageTableRef {
+    pub fn add_function(&self, data:VirtAddr, size: u64) {
+        let data: *mut u8 = data.as_mut_ptr::<u8>();
+        let data = unsafe { slice::from_raw_parts(data, size as usize) };
+        self.add_region_vaddr(VirtAddr::from(0xFE8000000000u64), data);
```
If I changed the address to, say, 
```rust
self.add_region_vaddr(VirtAddr::from(0x140_0000_0000u64), data);
```
then I confirmed that
```
[libos]
entrypoint = "/lib/python"

[loader]
argv = ["python3", "-c", "print('import ctypes'); import ctypes; print('string_at'); a=ctypes.string_at(0x14000000000); print(a)"]
...
```

this manifest works.

mmisono replied:
```
self.add_region_vaddr(VirtAddr::from(0x7F80_0000_0000u64), data);
```
is ok but
```
self.add_region_vaddr(VirtAddr::from(0x8000_0000_0000u64), data);
```
is not (PGD idx=256)

mmisono replied:
ok this is because
> Address sizes:                      43 bits physical, 48 bits virtual

mmisono commented on kernel/src/process_manager/process_paging.rs:
```diff
@@ -286,6 +286,12 @@ impl ProcessPageTableRef {
+    pub fn add_function(&self, data:VirtAddr, size: u64) {
+        let data: *mut u8 = data.as_mut_ptr::<u8>();
+        let data = unsafe { slice::from_raw_parts(data, size as usize) };
+        self.add_region_vaddr(VirtAddr::from(0xFE8000000000u64), data);
```
I suggest to use
```suggestion
        self.add_region_vaddr(VirtAddr::from(0x140_0000_0000u64), data);
```

mmisono commented on kernel/src/process_manager/memory_channels.rs:
```diff
@@ -0,0 +1,70 @@
+
+use super::{allocation::AllocationRange, process::ProcessID, process_paging::ProcessPageTableRef};
+
+pub const INPUT_VADDR: u64 = 0xFF0000000000u64;
+pub const OUTPUT_VADDR: u64 = 0xFF8000000000u64;
```
These exceed 48-bit virtual addresses. I suggest to use following instead
```suggestion
pub const INPUT_VADDR: u64 = 0x200_0000_0000u64;
pub const OUTPUT_VADDR: u64 = 0x200_0000_8000u64;
```

However, even with this change, I still cannot access input/output channels (it seems #PF)

mmisono commented on kernel/src/process_manager/allocation.rs:
```diff
@@ -7,39 +7,84 @@ use crate::cpu::control_regs::read_cr3;
+        let start_address = VirtAddr::from(start_addr);
 
         for i in 0..(pages as usize) {
             let current_page = allocate_page();
```
I think for non-mount case (i.e., adding pages for trustlets), rmpadjust is needed
```c
            if !mount {
                let (mapping, _page_mapped) = paddr_as_slice!(current_page);
                rmp_adjust(mapping.virt_addr(), RMPFlags::VMPL1 | RMPFlags::RWX , PageSize::Regular).unwra
p();
            }
```

mmisono commented on kernel/src/process_manager/memory_channels.rs:
```diff
@@ -0,0 +1,70 @@
+    pub fn copy_into(&mut self, source_addr: u64, page_table: u64, size: usize) {
+        let copy_size = size + PAGE_SIZE - (size % PAGE_SIZE);
+        let copy_page_count = copy_size / PAGE_SIZE;
+        let target = VirtAddr::from(ALLOCATION_RANGE_VIRT_START);
```
`INPUT_VADDR` istead of `ALLOCATION_RANGE_VIRT_STRAT`?

mmisono commented on kernel/src/process_manager/memory_channels.rs:
```diff
@@ -0,0 +1,70 @@
+
+        self.input.mount();
+
+        page_table_ref.copy_address_range(VirtAddr::from(source_addr), copy_size as u64, target);
```
In my understanding, this function is copying the guest input data to the input channel. For that I believe we need to copy the guest data first

mmisono commented on kernel/src/process_manager/memory_channels.rs:
```diff
@@ -0,0 +1,70 @@
+
+impl MemoryChannel {
+
+    pub fn allocate_input(&mut self, page_table_ref: &mut ProcessPageTableRef, size: usize) {
```
```
     pub fn allocate_input(&mut self, page_table_ref: &mut ProcessPageTableRef, size: usize) {
-        self.input = self.allocate_range(page_table_ref, size, INPUT_VADDR);
+        let flags = ProcessPageFlags::PRESENT | ProcessPageFlags::WRITABLE |
+        ProcessPageFlags::USER_ACCESSIBLE | ProcessPageFlags::ACCESSED;
+        page_table_ref.add_pages(VirtAddr::from(INPUT_VADDR), 1, flags);
+        //self.input = self.allocate_range(page_table_ref, size, INPUT_VADDR);
     }

     pub fn allocate_output(&mut self, page_table_ref: &mut ProcessPageTableRef, size: usize) {
-        self.output = self.allocate_range(page_table_ref, size, OUTPUT_VADDR);
+        let flags = ProcessPageFlags::PRESENT | ProcessPageFlags::WRITABLE |
+        ProcessPageFlags::USER_ACCESSIBLE | ProcessPageFlags::ACCESSED;
+        page_table_ref.add_pages(VirtAddr::from(OUTPUT_VADDR), 1, flags);
+        //self.output = self.allocate_range(page_table_ref, size, OUTPUT_VADDR);
     }
```

this works. allocate_range for the trustlets has some issuses

sabtf committed:
Fix Problems with memory channels

Set new ranges for input and output
Add rmp_adjust for new memory
Fix input copy
